### PR TITLE
feat(anvil): add eth_getProof implementation

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
* support `eth_getProof` rpc call

## Solution
* bindings
* implementation according to https://eips.ethereum.org/EIPS/eip-1186

Note: this is a bit inefficient because all our data is stored in `address -> Account` maps so we first have to encode the entire state and insert it in a `Trie`

* added a helper trait that can create a `HashDB` which is used by the `Trie` to write data into: `MaybeHashDatabase` this is a `maybe` to account for forking mode

this is a bit tricky to test without reference test cases, but this just uses the existing and tested trie implementations

### CODE REVIEW

File: anvil/src/eth/error.rs [hunk 1]
```diff
@@ -60,6 +60,10 @@ pub enum BlockchainError {
     BlockOutOfRange(u64, u64),
     #[error("Resource not found")]
     BlockNotFound,
+    #[error("Required data unavailable")]
+    DataUnavailable,
+    #[error("trie error: {0}")]
```
Comment (onbjerg) [inline anvil/src/eth/error.rs:65]:
```suggestion
    #[error("Trie error: {0}")]
```

**Review by onbjerg — APPROVED** [review]

**Comment by mattsse** [comment]
@gakonst pulled in a copy of unreleased https://github.com/paritytech/trie/tree/master/trie-eip1186 to verify proofs. 

pulling in via git causes some dep conflicts...

File: anvil/core/src/eth/proof.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    /// nonce of the account
+    pub nonce: U256,
+    /// SHA3 of the StorageRoot. All storage will deliver a MerkleProof starting with this
+    /// rootHash.
+    pub storage_hash: H256,
+    /// Array of storage-entries as requested. Each entry is a object with these properties:
+    pub storage_proof: Vec<StorageProof>,
+}
+
+/// Basic account type.
```
Comment (gakonst) [inline anvil/core/src/eth/proof.rs:41]: Maybe we should add these to revm? Or should we implement this as a newtype over the revm types?

**Review by gakonst — APPROVED** [review]

File: anvil/core/src/eth/proof.rs (hunk 2 above)
Comment (mattsse) [reply anvil/core/src/eth/proof.rs:41]: probably useful to have them in revm

**Pull request merged**